frame/staking/src/pallet/impls.rs

```diff
@@ -633,81 +643,136 @@ impl<T: Config> Pallet<T> {
 
 	/// Get all of the voters that are eligible for the npos election.
 	///
+	/// `voter_count` imposes a cap on the number of voters returned; care should be taken to ensure
+	/// that it is accurate.
+	///
 	/// This will use all on-chain nominators, and all the validators will inject a self vote.
 	///
 	/// ### Slashing
 	///
 	/// All nominations that have been submitted before the last non-zero slash of the validator are
 	/// auto-chilled.
-	///
-	/// Note that this is VERY expensive. Use with care.
-	pub fn get_npos_voters() -> Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)> {
-		let weight_of = Self::slashable_balance_of_fn();
-		let mut all_voters = Vec::new();
+	pub fn get_npos_voters(
+		maybe_max_len: Option<usize>,
+	) -> Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)> {
+		let weight_of = Self::weight_of_fn();
+
+		let nominator_count = CounterForNominators::<T>::get() as usize;
+		let validator_count = CounterForValidators::<T>::get() as usize;
+		let all_voter_count = validator_count.saturating_add(nominator_count);
+
+		let max_allowed_len = maybe_max_len.unwrap_or(all_voter_count).min(all_voter_count);
+		drop(all_voter_count);
 
-		for (validator, _) in <Validators<T>>::iter() {
+		let mut all_voters = Vec::<_>::with_capacity(max_allowed_len);
```

Comment:
I am confused how this `with_capacity` can possible work accurately given the rust runtimes knowledge of the variable length of items. In the case of nominators we are adding the items `(AccountId, VoteWeight, Vec<AccountId>)`. AFAIU `Vec<AccountId>` could be _any_ length. 

I wonder if there is some optimization here where we use a macro that generates fixed length arrays of length 1 -> MaxNominationTargets. Then when we alloc for this in the worst case the rust runtime would know that the biggest possible item is `(AccountId, VoteWeight, [AccountId; 16])` and allocate accordingly. (a lot of handwaving on my part here bc I don't actually know how rust/wasm does allocs)

Reply:
This is great question. 

While I don't type in the exact type of the vector, nonetheless rustc will know at this point that it will be `(AccountId, VoteWeight, Vec<AccountId>)`. The size of this tuple is fixed and known, and therefore the `with_capacity` can work. Your main mistake is that you think the nested `Vec<AccountId>` is stored in the same memory slab, while this is not the case. The nested `Vec<_>` is only stored as a pointer (to its actual data in the heap) in the top level tuple, which is stored in the stack. 

Everything that we've spoken about so far happens on the stack. You *never* store any dynamically sized data on the stake. In cases like the `Vec<_>`, you only allocate enough space for the pointer of the vector on the stack, the data itself is in the heap.

As a final note, all data types in rust have a known size (on the stack) at compile time (except those that are `?Sized`). For vec, it is just a pointer (and maybe the length, not sure) that is stored in the stakc. Most `std::collections` are similar. Also, `Box<T>` means "move T to the heap and keep a pointer to it in the stack". 

More detail: https://doc.rust-lang.org/1.22.0/book/first-edition/the-stack-and-the-heap.html